Build error on type checking on the llvm-12 branch

The build for llvm-hs-pretty on the llvm12 branch fails (I have tried a few different commits).  The error is:

llvm-hs-pretty> configure
llvm-hs-pretty> Configuring llvm-hs-pretty-12.0.0...
llvm-hs-pretty> build
llvm-hs-pretty> Preprocessing library for llvm-hs-pretty-12.0.0..
llvm-hs-pretty> Building library for llvm-hs-pretty-12.0.0..
llvm-hs-pretty> [1 of 1] Compiling LLVM.Pretty
llvm-hs-pretty>
llvm-hs-pretty> /tmp/stack-08f048299826d3a9/llvm-hs-pretty-12.0.0/src/LLVM/Pretty.hs:538:11: error:
llvm-hs-pretty>     • Couldn't match expected type ‘m7 (Either String Type)’
llvm-hs-pretty>                   with actual type ‘Type’
llvm-hs-pretty>     • In the pattern: PointerType argTy_ _
llvm-hs-pretty>       In a case alternative: PointerType argTy_ _ -> argTy_
llvm-hs-pretty>       In the expression:
llvm-hs-pretty>         case typeOf address of
llvm-hs-pretty>           PointerType argTy_ _ -> argTy_
llvm-hs-pretty>           _ -> error "invalid load of non-pointer type. (Malformed AST)"
llvm-hs-pretty>     |
llvm-hs-pretty> 538 |           PointerType argTy_ _ -> argTy_
llvm-hs-pretty>     |           ^^^^^^^^^^^^^^^^^^^^
llvm-hs-pretty>
llvm-hs-pretty> /tmp/stack-08f048299826d3a9/llvm-hs-pretty-12.0.0/src/LLVM/Pretty.hs:553:38: error:
llvm-hs-pretty>     • Couldn't match expected type ‘Type’
llvm-hs-pretty>                   with actual type ‘m8 (Either String Type)’
llvm-hs-pretty>     • In the second argument of ‘($)’, namely ‘typeOf address’
llvm-hs-pretty>       In the expression: getElementType $ typeOf address
llvm-hs-pretty>       In an equation for ‘argTy’: argTy = getElementType $ typeOf address
llvm-hs-pretty>     |
llvm-hs-pretty> 553 |       where argTy = getElementType $ typeOf address
llvm-hs-pretty>     |                                      ^^^^^^^^^^^^^^
llvm-hs-pretty>
llvm-hs-pretty> /tmp/stack-08f048299826d3a9/llvm-hs-pretty-12.0.0/src/LLVM/Pretty.hs:1133:9: error:
llvm-hs-pretty>     • Couldn't match expected type ‘m6 (Either String Type)’
llvm-hs-pretty>                   with actual type ‘Type’
llvm-hs-pretty>     • In the pattern: PointerType argTy_ _
llvm-hs-pretty>       In a case alternative: PointerType argTy_ _ -> argTy_
llvm-hs-pretty>       In the expression:
llvm-hs-pretty>         case typeOf address of
llvm-hs-pretty>           PointerType argTy_ _ -> argTy_
llvm-hs-pretty>           _ -> error "invalid load of non-pointer type. (Malformed AST)"
llvm-hs-pretty>      |
llvm-hs-pretty> 1133 |         PointerType argTy_ _ -> argTy_
llvm-hs-pretty>      |         ^^^^^^^^^^^^^^^^^^^^
llvm-hs-pretty>
llvm-hs-pretty> /tmp/stack-08f048299826d3a9/llvm-hs-pretty-12.0.0/src/LLVM/Pretty.hs:1319:44: error:
llvm-hs-pretty>     • Couldn't match expected type ‘Type’
llvm-hs-pretty>                   with actual type ‘m4 (Either String Type)’
llvm-hs-pretty>     • In the first argument of ‘referencedType’, namely ‘(typeOf f)’
llvm-hs-pretty>       In the expression: (referencedType (typeOf f))
llvm-hs-pretty>       In the expression:
llvm-hs-pretty>         case (referencedType (typeOf f)) of
llvm-hs-pretty>           fty@FunctionType {..} -> fty
llvm-hs-pretty>           _ -> error "Calling non function type. (Malformed AST)"
llvm-hs-pretty>      |
llvm-hs-pretty> 1319 |       functionType = case (referencedType (typeOf f)) of
llvm-hs-pretty>      |                                            ^^^^^^^^
llvm-hs-pretty>
llvm-hs-pretty> /tmp/stack-08f048299826d3a9/llvm-hs-pretty-12.0.0/src/LLVM/Pretty.hs:1362:43: error:
llvm-hs-pretty>     • Couldn't match expected type ‘Type’
llvm-hs-pretty>                   with actual type ‘m5 (Either String Type)’
llvm-hs-pretty>     • In the first argument of ‘referencedType’, namely ‘(typeOf f)’
llvm-hs-pretty>       In the expression: referencedType (typeOf f)
llvm-hs-pretty>       In the expression:
llvm-hs-pretty>         case referencedType (typeOf f) of
llvm-hs-pretty>           fty@FunctionType {..} -> fty
llvm-hs-pretty>           _ -> error "Invoking non-function type. (Malformed AST)"
llvm-hs-pretty>      |
llvm-hs-pretty> 1362 |       functionType = case referencedType (typeOf f) of
llvm-hs-pretty>      |                                           ^^^^^^^^
llvm-hs-pretty>
Progress 1/2

--  While building package llvm-hs-pretty-12.0.0 (scroll up to its section to see the error) using:
      /home/jdgallag/.stack/setup-exe-cache/x86_64-linux-tinfo6/Cabal-simple_mPHDZzAJ_3.2.1.0_ghc-8.10.4 --builddir=.stack-work/dist/x86_64-linux-tinfo6/Cabal-3.2.1.0 build --ghc-options " -fdiagnostics-color=always"
    Process exited with code: ExitFailure 1


In the Stack.yaml file, I have 

extra-deps:
  (...)
  - git: https://github.com/llvm-hs/llvm-hs-pretty.git
    commit: c26764e2dcef78d640bbbc18518eaabda92603c0
(this isn't the latest commit here, but the issue seems to exist on multiple commits, at least for me).

And in the Package.yaml file I have 
dependencies:
  (...)
- llvm-hs-pretty

I also tried without and without combinations of the following options in Stack.yaml:

ghc-options:
  llvm-hs: -optcxx=-std=c++14 -optcxx=-lstdc++ -optcxx=-fno-rtti


flags:
  llvm-hs:
    shared-llvm: true


Is the build indeed failing, or am I missing something obvious here.  The llvm-hs and llvm-hs-pure packages both succeed on the llvm-12 branch.

--- a/llvm-hs-pretty.cabal
+++ b/llvm-hs-pretty.cabal
@@ -1,7 +1,7 @@
 name:                llvm-hs-pretty
 version:             0.9.0.0
 license:             MIT
-synopsis:            A pretty printer for LLVM IR. 
+synopsis:            A pretty printer for LLVM IR.
 description:         A pure Haskell pretty printer for the LLVM AST types provided by llvm-hs.
 license-file:        LICENSE
 author:              Stephen Diehl
@@ -18,8 +18,9 @@ Source-Repository head
 
 library
   hs-source-dirs:      src
-  exposed-modules:     
-    LLVM.Pretty
+  exposed-modules:
+    LLVM.Pretty,
+    LLVM.Pretty.Typed
   ghc-options:
     -fwarn-incomplete-patterns
   default-language:    Haskell2010
@@ -36,7 +37,7 @@ Test-suite test
   hs-source-dirs:      tests
   main-is:             Main.hs
   default-language:    Haskell2010
-  build-depends: 
+  build-depends:
     llvm-hs-pretty,
     base                 >= 4.6  && < 5.0,
     text                 >= 0.1  && < 1.3,

--- a/src/LLVM/Pretty.hs
+++ b/src/LLVM/Pretty.hs
@@ -17,7 +17,7 @@ module LLVM.Pretty (
 import Prelude hiding ((<$>))
 import GHC.Word
 
-import LLVM.AST.Typed
+import LLVM.Pretty.Typed
 
 import LLVM.AST
 import LLVM.AST.Global

--- a/src/LLVM/Pretty/Typed.hs
+++ b/src/LLVM/Pretty/Typed.hs
@@ -1,0 +1,137 @@
+{-# LANGUAGE RecordWildCards, FlexibleInstances #-}
+
+-- TODO: remove this copy of the old "typeOf" implementation in llvm-hs-pure.
+-- It's needed because the new type signature of "typeOf" has a Monad constraint,
+-- which is not compatible with the pretty printing library that is used.
+
+-- | Querying the type of LLVM expressions
+module LLVM.Pretty.Typed (
+  Typed(..),
+  getElementType,
+  getElementPtrType,
+  extractValueType,
+) where
+
+import LLVM.Prelude
+
+import LLVM.AST
+import LLVM.AST.Global
+import LLVM.AST.Type
+
+import qualified LLVM.AST.Constant as C
+import qualified LLVM.AST.Float as F
+
+class Typed a where
+    typeOf :: a -> Type
+
+instance Typed Operand where
+  typeOf (LocalReference t _) = t
+  typeOf (ConstantOperand c)  = typeOf c
+  typeOf _                    = MetadataType
+
+instance Typed CallableOperand where
+  typeOf (Right op) = typeOf op
+  typeOf (Left _) = error "typeOf inline assembler is not defined. (Malformed AST)"
+
+instance Typed C.Constant where
+  typeOf (C.Int bits _)  = IntegerType bits
+  typeOf (C.Float t) = typeOf t
+  typeOf (C.Null t)      = t
+  typeOf (C.AggregateZero t) = t
+  typeOf (C.Struct {..}) = StructureType isPacked (map typeOf memberValues)
+  typeOf (C.Array {..})  = ArrayType (fromIntegral $ length memberValues) memberType
+  typeOf (C.Vector {..}) = VectorType (fromIntegral $ length memberValues) $
+                              case memberValues of
+                                  []    -> error "Vectors of size zero are not allowed. (Malformed AST)"
+                                  (x:_) -> typeOf x
+  typeOf (C.Undef t)     = t
+  typeOf (C.BlockAddress {..})   = ptr i8
+  typeOf (C.GlobalReference t _) = t
+  typeOf (C.Add {..})     = typeOf operand0
+  typeOf (C.FAdd {..})    = typeOf operand0
+  typeOf (C.FDiv {..})    = typeOf operand0
+  typeOf (C.FRem {..})    = typeOf operand0
+  typeOf (C.Sub {..})     = typeOf operand0
+  typeOf (C.FSub {..})    = typeOf operand0
+  typeOf (C.Mul {..})     = typeOf operand0
+  typeOf (C.FMul {..})    = typeOf operand0
+  typeOf (C.UDiv {..})    = typeOf operand0
+  typeOf (C.SDiv {..})    = typeOf operand0
+  typeOf (C.URem {..})    = typeOf operand0
+  typeOf (C.SRem {..})    = typeOf operand0
+  typeOf (C.Shl {..})     = typeOf operand0
+  typeOf (C.LShr {..})    = typeOf operand0
+  typeOf (C.AShr {..})    = typeOf operand0
+  typeOf (C.And {..})     = typeOf operand0
+  typeOf (C.Or  {..})     = typeOf operand0
+  typeOf (C.Xor {..})     = typeOf operand0
+  typeOf (C.GetElementPtr {..}) = getElementPtrType (typeOf address) indices
+  typeOf (C.Trunc {..})   = type'
+  typeOf (C.ZExt {..})    = type'
+  typeOf (C.SExt {..})    = type'
+  typeOf (C.FPToUI {..})  = type'
+  typeOf (C.FPToSI {..})  = type'
+  typeOf (C.UIToFP {..})  = type'
+  typeOf (C.SIToFP {..})  = type'
+  typeOf (C.FPTrunc {..}) = type'
+  typeOf (C.FPExt {..})   = type'
+  typeOf (C.PtrToInt {..}) = type'
+  typeOf (C.IntToPtr {..}) = type'
+  typeOf (C.BitCast {..})  = type'
+  typeOf (C.ICmp {..})    = case (typeOf operand0) of
+                              (VectorType n _) -> VectorType n i1
+                              _ -> i1
+  typeOf (C.FCmp {..})    = case (typeOf operand0) of
+                              (VectorType n _) -> VectorType n i1
+                              _ -> i1
+  typeOf (C.Select {..})  = typeOf trueValue
+  typeOf (C.ExtractElement {..})  = case typeOf vector of
+                                      (VectorType _ t) -> t
+                                      _ -> error "The first operand of an extractelement instruction is a value of vector type. (Malformed AST)"
+  typeOf (C.InsertElement {..})   = typeOf vector
+  typeOf (C.ShuffleVector {..})   = case (typeOf operand0, typeOf mask) of
+                                      (VectorType _ t, VectorType m _) -> VectorType m t
+                                      _ -> error "The first operand of an shufflevector instruction is a value of vector type. (Malformed AST)"
+  typeOf (C.ExtractValue {..})    = extractValueType indices' (typeOf aggregate)
+  typeOf (C.InsertValue {..})     = typeOf aggregate
+  typeOf (C.TokenNone)          = TokenType
+  typeOf (C.AddrSpaceCast {..}) = type'
+
+getElementPtrType :: Type -> [C.Constant] -> Type
+getElementPtrType ty [] = ptr ty
+getElementPtrType (PointerType ty _) (_:is) = getElementPtrType ty is
+getElementPtrType (StructureType _ elTys) (C.Int 32 val:is) =
+  getElementPtrType (elTys !! fromIntegral val) is
+getElementPtrType (VectorType _ elTy) (_:is) = getElementPtrType elTy is
+getElementPtrType (ArrayType _ elTy) (_:is) = getElementPtrType elTy is
+getElementPtrType _ _ = error "Expecting aggregate type. (Malformed AST)"
+
+getElementType :: Type -> Type
+getElementType (PointerType t _) = t
+getElementType _ = error $ "Expecting pointer type. (Malformed AST)"
+
+extractValueType :: [Word32] -> Type -> Type
+extractValueType [] ty = ty
+extractValueType (i : is) (ArrayType numEls elTy)
+  | fromIntegral i < numEls = extractValueType is elTy
+  | fromIntegral i >= numEls = error "Expecting valid index into array type. (Malformed AST)"
+extractValueType (i : is) (StructureType _ elTys)
+  | fromIntegral i < length elTys = extractValueType is (elTys !! fromIntegral i)
+  | otherwise = error "Expecting valid index into structure type. (Malformed AST)"
+extractValueType _ _ = error "Expecting vector type. (Malformed AST)"
+
+instance Typed F.SomeFloat where
+  typeOf (F.Half _)          = FloatingPointType HalfFP
+  typeOf (F.Single _)        = FloatingPointType FloatFP
+  typeOf (F.Double _)        = FloatingPointType DoubleFP
+  typeOf (F.Quadruple _ _)   = FloatingPointType FP128FP
+  typeOf (F.X86_FP80 _ _)    = FloatingPointType X86_FP80FP
+  typeOf (F.PPC_FP128 _ _)   = FloatingPointType PPC_FP128FP
+
+instance Typed Global where
+  typeOf (GlobalVariable {..}) = type'
+  typeOf (GlobalAlias {..})    = type'
+  typeOf (Function {..})       = let (params, isVarArg) = parameters
+                                   in FunctionType returnType (map typeOf params) isVarArg
+instance Typed Parameter where
+  typeOf (Parameter t _ _) = t

